spark-api-jupyter/templates/deployment.yaml

```diff
@@ -0,0 +1,62 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "fullname" . }}
+  labels:
+    app: {{ template "name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
+  replicas: {{ .Values.replicaCount }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "name" . }}
+        release: {{ .Release.Name }}
+    spec:
+      containers:
+        - name: {{ .Chart.Name }}
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          {{ if .Values.jupyter.password }}
+          command: ["start-notebook.sh", "--NotebookApp.password={{ .Values.jupyter.password }}"]
+          {{ end }}
+          ports:
+            - containerPort: {{ .Values.service.internalPort }}
+          livenessProbe:
+            httpGet:
+              path: /
+              port: {{ .Values.service.internalPort }}
+          readinessProbe:
+            httpGet:
+              path: /
+              port: {{ .Values.service.internalPort }}
+          resources:
+{{ toYaml .Values.resources | indent 12 }}
+          volumeMounts:
+            - mountPath: {{ .Values.jupyter.repositoriesPath }}
+              name: {{ .Values.jupyter.repositoriesDbDisk }}
+              readOnly: true
+            - mountPath: {{ .Values.jupyter.userPath }}
+              name: {{ .Values.jupyter.userDbDisk }}
+
+    # to ease the pain of writing this, especially from terraform
+    {{- if .Values.hostnameNodeSelector }}
+      nodeSelector:
+        kubernetes.io/hostname: {{ .Values.hostnameNodeSelector }}
+    {{- end }}
+    {{- if .Values.nodeSelector }}
```

Comment:
Why `.Values.nodeSelector` _and_ `.Values.hostnameNodeSelector`? How are they different?

Reply:
`nodeSelector` is a general selector whereas `hostnameNodeSelector` adds specifically a `kubernetes.io/hostname` selector. Mainly two reasons

1. `--set` syntax is extremely clumsy for complex selectors in the general case

   ```
   nodeSelector.kubernetes\.io/hostname=gke-spark-api-default-pool-021cf524-sksl
   ```
2. That didn't allow me to have an empty hostname selector in terraform

As expected, the more general selector overrides